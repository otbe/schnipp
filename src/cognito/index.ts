import { CognitoUserPoolTriggerEvent } from 'aws-lambda';
import { IHandler } from '../';

export type CognitoUserPoolHandler = IHandler<
  CognitoUserPoolTriggerEvent,
  CognitoUserPoolTriggerEvent
>;
