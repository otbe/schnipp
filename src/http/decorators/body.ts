import { createMethodParameterDecorator } from '../factories/createMethodParamDecorator';

export const Body = () =>
  createMethodParameterDecorator(e => JSON.parse(e.body || '{}'));
