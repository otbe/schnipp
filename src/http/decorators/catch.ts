import { inject } from 'simple-ts-di';

const CATCH = Symbol('catch');

export const Catch = (...exceptions: any[]): ClassDecorator => (
  target: any
) => {
  inject()(target);
  Reflect.defineMetadata(CATCH, exceptions, target.constructor);
};

export const getCaughtExceptions = (catcher: any) =>
  Reflect.getMetadata(CATCH, catcher.constructor) || [];
