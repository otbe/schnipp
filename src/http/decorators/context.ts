import { createMethodParameterDecorator } from '../factories/createMethodParamDecorator';

export const Context = () =>
  createMethodParameterDecorator((e, m, context) => context);
