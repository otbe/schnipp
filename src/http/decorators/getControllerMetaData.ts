import { DecoratedExceptionFilter, Guard, MetaData } from '..';
import { ContainedType } from '../../container';
import { RestController } from '../RestController';

export const CONTROLLER = Symbol('controller');

export type ControllerBaseData = {
  headers: { [key: string]: string };
  filters: Array<DecoratedExceptionFilter>;
  guards: Array<ContainedType<Guard>>;
  metaData: MetaData;
};

export type RestControllerData = ControllerBaseData & {
  basePath?: string;
};

export type ControllerData = RestControllerData;

export const getControllerMetaData = (controller: typeof RestController) => {
  let data: ControllerData | null = Reflect.getMetadata(CONTROLLER, controller);

  if (data == null) {
    data = {
      filters: [],
      guards: [],
      metaData: {},
      headers: {}
    };

    Reflect.defineMetadata(CONTROLLER, data, controller);
  }

  return data;
};
