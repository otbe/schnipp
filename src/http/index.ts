import {
  APIGatewayEvent,
  APIGatewayProxyResult,
  CustomAuthorizerEvent,
  CustomAuthorizerResult
} from 'aws-lambda';
import { IHandler } from '..';

export type APIGatewayHandler = IHandler<
  APIGatewayEvent,
  APIGatewayProxyResult
>;

export type AuthorizeHandler = IHandler<
  CustomAuthorizerEvent,
  CustomAuthorizerResult
>;

export { AuthPolicy, HttpVerb, Effect } from './utils/AuthPolicy';
export { createResponse } from './utils/createResponse';
export { MetaData } from './utils/MetaData';
export {
  DecoratedExceptionFilter,
  ExceptionFilter
} from './utils/ExceptionFilter';
export { Guard } from './utils/Guard';
export { Catch } from './decorators/catch';
export { RestController } from './RestController';

export {
  InjectedParam,
  MethodMetaData,
  ParamInjector
} from './factories/createHttpMethodDecorator';

export {
  GET,
  POST,
  DELETE,
  PUT,
  PATCH,
  OPTIONS,
  HEAD
} from './decorators/methods';

export { Path } from './decorators/path';
export { Param } from './decorators/param';
export { Query } from './decorators/query';
export { StatusCode } from './decorators/statusCode';
export { Context } from './decorators/context';
export { Body } from './decorators/body';
export { Header } from './decorators/header';
export { Meta } from './decorators/meta';
export { Use } from './decorators/use';

export { HttpException } from './exceptions/HttpException';
export { NotFoundException } from './exceptions/NotFoundException';
export { ForbiddenException } from './exceptions/ForbiddenException';
