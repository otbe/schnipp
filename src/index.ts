import { Context, Handler } from 'aws-lambda';
import { ContainedType, getFromContainer } from './container';

export type IHandler<E, R = void> = {
  handle(event: E, context: Context): Promise<R> | R;
};

let setupPromise: Promise<any> | null = null;
export const createHandler = <E, R>(
  handler: ContainedType<IHandler<E, R>>,
  setupFn?: () => Promise<any>
): Handler<E, R> => async (event: E, context: Context) => {
  context.callbackWaitsForEmptyEventLoop = false;

  if (setupFn) {
    if (setupPromise == null) {
      setupPromise = setupFn();
    }

    await setupPromise;
  }

  const handlerInstance = getFromContainer(handler);
  return await handlerInstance.handle(event, context);
};

export { useContainer } from './container';
