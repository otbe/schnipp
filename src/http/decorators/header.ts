import { getControllerMetaData } from './getControllerMetaData';
import { getMethodMetaData } from '../factories/createHttpMethodDecorator';

export const Header = (name: string, value: string) => (
  ...args: Array<any>
) => {
  const target = args[0];

  switch (args.length) {
    // class
    case 1:
      const controllerMetaData = getControllerMetaData(target);
      controllerMetaData.headers[name] = value;
      return;
    // method
    case 2:
    case 3:
      const methodMetaData = getMethodMetaData(args[1], target);
      methodMetaData.headers[name] = value;
      return args[2];
    default:
      throw 'not supported target for @Header';
  }
};
