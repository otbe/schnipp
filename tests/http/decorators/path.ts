import 'reflect-metadata';
import { Path, RestController } from '../../../src/http';
import { getControllerMetaData } from '../../../src/http/decorators/getControllerMetaData';

describe('@Path', () => {
  it('should add metadata for controller', () => {
    @Path('/bar')
    class Test extends RestController {}

    expect(getControllerMetaData(Test)).toMatchSnapshot();
  });
});
