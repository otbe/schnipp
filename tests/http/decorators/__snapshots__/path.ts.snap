// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Path should add metadata for controller 1`] = `
Object {
  "basePath": "/bar",
  "filters": Array [],
  "guards": Array [],
  "headers": Object {},
  "metaData": Object {},
}
`;
